pkg/apis/migration/migration.go

```diff
@@ -0,0 +1,351 @@
+/*
+Copyright 2020 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package migration
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"reflect"
+	"regexp"
+	"strings"
+	"sync"
+	"time"
+
+	"github.com/davecgh/go-spew/spew"
+	"github.com/google/uuid"
+	cnstypes "github.com/vmware/govmomi/cns/types"
+	apiextensionsv1beta1 "k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/v1beta1"
+	apiextensionsclient "k8s.io/apiextensions-apiserver/pkg/client/clientset/clientset"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/util/wait"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/client/config"
+
+	migration "sigs.k8s.io/vsphere-csi-driver/pkg/apis/migration/v1alpha1"
+	"sigs.k8s.io/vsphere-csi-driver/pkg/common/cns-lib/vsphere"
+	"sigs.k8s.io/vsphere-csi-driver/pkg/csi/service/common"
+	"sigs.k8s.io/vsphere-csi-driver/pkg/csi/service/logger"
+	k8s "sigs.k8s.io/vsphere-csi-driver/pkg/kubernetes"
+)
+
+const (
+	timeout  = 60 * time.Second
+	pollTime = 5 * time.Second
+)
+
+// ErrVolumeNotRegistered is returned when requested volumepath is not found from cache and K8S CR List
+var ErrVolumeNotRegistered = errors.New("volume is not registered to CNS")
+
+// VolumeMigrationService exposes interfaces to manage VolumeInfo cache and CR
+// for VolumePath, VolumeID and VolumeName mapping
+type VolumeMigrationService interface {
```

Comment:
We only need `GetVolumeInfo` and `DeleteVolumeInfo`. `GetVolumeInfo` can register the volume if not registered, save CR etc. And VolumeMigrationService init can load all volume info internally.

Reply:
There are some use cases for `GetVolumeInfo()` without registering volume in the syncer/fullsync container.

When PV/PVC gets updated we want to push metadata only if volume is fully migrated to CSI (attach is completed by CSI driver and not just annotated with migrated). For this reason I kept separate interfaces `GetVolumeInfo()`, `RegisterVolume()` and `SaveVolumeInfo()`.

Reply:
Can we discuss this scenario?

Reply:
as discussed kept three methods `GetVolumeID`, `GetVolumePath` and `DeleteVolumeInfo`  which cover all cases.

Hide all cache management and CR related logic behind, so the caller does not need to call them explicitly.


```
// VolumeMigrationService exposes interfaces to manage VolumeInfo cache and CR
// for VolumePath, VolumeID and VolumeName mapping
type VolumeMigrationService interface {
	// GetVolumeID returns VolumeID for given VolumePath from local cache
	// if mapping is not present in the cache, then VolumeID is retrieved by calling CNS CreateVolume with setting BackingDiskURLPath
	// After VolumeID is retrieved from vCenter, CR will be created to persist mapping and cache will be updated.
	// Returns an error if any operation fails.
	GetVolumeID(ctx context.Context, volumePath string) (string, error)

	// GetVolumePath returns volumePath for given volumeID from local cache
	// if mapping is not available in the cache, QueryVolumeInfo will be called
	// to retrieve volumePath for given VolumeID
	GetVolumePath(ctx context.Context, volumeID string) (string, error)

	// DeleteVolumeInfo helps delete mapping of volumePath to VolumeID for specified volumeID
	DeleteVolumeInfo(ctx context.Context, volumeID string) error
}
```